File: source/qdk-mcp/test-server.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

"""Validate the qdk-mcp server over stdio JSON-RPC."""

import json
import subprocess
import sys
import threading


def main():
    server = subprocess.Popen(
        [sys.executable, "-m", "qdk_mcp.server"],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )

    # Forward stderr in a background thread
    def forward_stderr():
        for line in server.stderr:
            sys.stderr.buffer.write(line)
            sys.stderr.buffer.flush()

    threading.Thread(target=forward_stderr, daemon=True).start()

    msg_id = 0

    def send(method, params):
        nonlocal msg_id
        msg_id += 1
        msg = json.dumps(
            {"jsonrpc": "2.0", "id": msg_id, "method": method, "params": params}
        )
        server.stdin.write((msg + "\n").encode())
        server.stdin.flush()
        return msg_id

    def recv():
        line = server.stdout.readline().decode().strip()
        if not line:
            raise RuntimeError("Server closed stdout unexpectedly")
        return json.loads(line)

    try:
        # 1. Initialize
        send(
            "initialize",
            {
                "protocolVersion": "2025-03-26",
                "capabilities": {},
                "clientInfo": {"name": "test", "version": "1.0"},
            },
        )
        data = recv()
        assert "result" in data, f"Expected result in init response: {data}"
        info = data["result"]["serverInfo"]
        print("=== INIT ===")
        print(f"Server: {info['name']}")
        print(f"Capabilities: {json.dumps(data['result']['capabilities'])}")
        print()

        # 2. List tools
        send("tools/list", {})
        data = recv()
        tools = data["result"]["tools"]
        print("=== TOOLS ===")
        tool_names = set()
        for t in tools:
            tool_names.add(t["name"])
            print(f"- {t['name']}: {t.get('description', '')[:80]}")
            meta_ui = (t.get("_meta") or {}).get("ui")
            if meta_ui:
                print(f"  UI: {json.dumps(meta_ui)}")
        print()

        assert "eval" in tool_names, "Missing 'eval' tool"
        assert "circuit" in tool_names, "Missing 'circuit' tool"

        # 3. List resources
        send("resources/list", {})
        data = recv()
        resources = data["result"]["resources"]
        print("=== RESOURCES ===")
        for r in resources:
            print(f"- {r['uri']} ({r.get('mimeType', 'N/A')})")
        print()

        assert any(
            r["uri"] == "ui://qdk/circuit.html" for r in resources
        ), "Missing circuit.html resource"

        # 4. Call eval tool
        send(
            "tools/call",
            {
                "name": "eval",
                "arguments": {
                    "source": "1 + 1",
                },
            },
        )
        data = recv()
        print("=== EVAL RESULT ===")
        assert "result" in data, f"Expected result: {data}"
        eval_text = data["result"]["content"][0]["text"]
        eval_parsed = json.loads(eval_text)
        print(f"Result: {eval_parsed['result']}")
        assert eval_parsed["result"] == 2, f"Expected 2, got {eval_parsed['result']}"
        print()

        # 5. Call circuit tool
        send(
            "tools/call",
            {
                "name": "circuit",
                "arguments": {
                    "entry_expr": "{ use q = Qubit(); H(q); let r = M(q); Reset(q); r }",
                },
            },
        )
        data = recv()
        print("=== CIRCUIT RESULT ===")
        assert "result" in data, f"Expected result: {data}"
        result = data["result"]
        assert not result.get("isError"), f"Circuit call returned error: {result}"
        print(f"Text: {result['content'][0]['text'][:200]}...")
        sc = result.get("structuredContent")
        assert sc is not None, "Expected structuredContent in circuit result"
        assert "circuits" in sc, "Expected 'circuits' key in structuredContent"
        assert "version" in sc, "Expected 'version' key in structuredContent"
        assert len(sc["circuits"]) >= 1, "Expected at least 1 circuit"
        circ = sc["circuits"][0]
        qubits = circ.get("qubits", [])
        grid = circ.get("componentGrid", [])
        print(f"Qubits: {len(qubits)}")
        print(f"Gate columns: {len(grid)}")
        assert len(qubits) >= 1, f"Expected at least 1 qubit, got {len(qubits)}"
        assert len(grid) >= 1, f"Expected at least 1 gate column, got {len(grid)}"
        print()

        # 6. Call circuit tool with invalid generation_method
        send(
            "tools/call",
            {
                "name": "circuit",
                "arguments": {
                    "entry_expr": "{ use q = Qubit(); H(q); }",
                    "generation_method": "InvalidMethod",
                },
            },
        )
        data = recv()
        print("=== CIRCUIT ERROR TEST ===")
        result = data.get("result", {})
        assert result.get(
            "isError"
        ), "Expected isError=True for invalid generation_method"
        error_text = result["content"][0]["text"]
        print(f"Error (expected): {error_text}")
        assert "InvalidMethod" in error_text
        print()

        print("=== ALL TESTS PASSED ===")

    finally:
        server.kill()
        server.wait()


if __name__ == "__main__":
    main()

File: source/qdk-mcp-ts/test-server.mjs
import { spawn } from "node:child_process";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const __dirname = dirname(fileURLToPath(import.meta.url));

const server = spawn("node", [join(__dirname, "dist/index.js"), "--stdio"], {
  stdio: ["pipe", "pipe", "pipe"],
});

let buffer = "";
let id = 0;

function send(method, params) {
  const msg = JSON.stringify({ jsonrpc: "2.0", id: ++id, method, params });
  server.stdin.write(msg + "\n");
  return id;
}

function onMessage(msg) {
  const data = JSON.parse(msg);
  if (data.id === 1) {
    console.log("=== INIT ===");
    console.log(
      "Server:",
      data.result.serverInfo.name,
      data.result.serverInfo.version,
    );
    console.log("Capabilities:", JSON.stringify(data.result.capabilities));
    console.log();
    send("tools/list", {});
  } else if (data.id === 2) {
    console.log("=== TOOLS ===");
    for (const t of data.result.tools) {
      console.log(`- ${t.name}: ${t.description}`);
      if (t._meta?.ui) console.log("  UI:", JSON.stringify(t._meta.ui));
    }
    console.log();
    send("resources/list", {});
  } else if (data.id === 3) {
    console.log("=== RESOURCES ===");
    for (const r of data.result.resources) {
      console.log(`- ${r.uri} (${r.mimeType})`);
    }
    console.log();
    send("tools/call", {
      name: "circuit",
      arguments: {
        source: `operation Main() : Result { use q = Qubit(); H(q); let r = M(q); Reset(q); r }`,
      },
    });
  } else if (data.id === 4) {
    console.log("=== CIRCUIT RESULT ===");
    if (data.error) {
      console.log("ERROR:", JSON.stringify(data.error, null, 2));
    } else {
      console.log("Text:", data.result.content[0].text);
      const sc = data.result.structuredContent;
      if (sc) {
        console.log("Structured content version:", sc.version);
        console.log("Circuits:", sc.circuits?.length);
        const c = sc.circuits?.[0];
        if (c) {
          console.log("Qubits:", c.qubits?.length);
          console.log("Gate columns:", c.componentGrid?.length);
        }
      }
    }
    console.log("\n=== ALL TESTS PASSED ===");
    server.kill();
  }
}

server.stdout.on("data", (chunk) => {
  buffer += chunk.toString();
  const lines = buffer.split("\n");
  buffer = lines.pop();
  for (const line of lines) {
    if (line.trim()) onMessage(line.trim());
  }
});

server.stderr.on("data", (chunk) => {
  process.stderr.write(chunk);
});

server.on("close", (code) => {
  process.exit(code ?? 0);
});

// Kick it off
send("initialize", {
  protocolVersion: "2025-03-26",
  capabilities: {},
  clientInfo: { name: "test", version: "1.0" },
});
